### PR TITLE
Some stability and logging updates as well as a new feature to get container metrics during cache refresh

### PR DESCRIPTION
**Pull request by jonathannewell**
Hi,

First, let me say thanks! Firehose-to-syslog is a cool tool. Second, let me say this is my first time writing go and Im hoping I didn't stray to far from accepted practice. So, whatever feedback you can offer me in that area would be greatly appreciated. Ok, to get to the point...I have been experiencing some stability issues with firehose-to-syslog in my environment. While they may only be related to my CF deployment it seemed like a few things in the tool would make it more robust. In order to help me determine what exactly was going on I also modified/adding logging in places. I also added a new cmd line argument and feature to pull container metrics during app caching. 

The issues/changes are:

* for stability I added deferred recover calls in several places throughout the app. Hopefully this isn't bad form.

* firehose-to-syslog would start up and throw an EOF on the firehose chan(no explanation) after about 10 minutes. Once this occurs once or twice (the executable doesn’t abend) I would stop getting logs in Logstash. It will throw 5 of those EOFs before failing completely. (Note: the subscription/connection being made with the NOAA client has 5 retries built-in.)

* As my uaa token would expire it would seem to auto-refresh a few times but eventually I would get a not-authorized error and the executable would abend.

* The caching function seemed to cause panics and abend the app intermittantly. I finally tracked this down to an App Guid that would come through not as a UUID or nil but as a string "\<nil\>". I modified that logic a little in the AnnotateWithAppData func in events to solve this. Once again, this may be something that was caused by my environment but it can't hurt to add the extra check. _Note: this is our lab environment and we currently only have one app deployed there._

* As I stated above I was never seeing any ‘Container Metric Events’ come out of the firehose. After some investigation I found a separate function that returned them in the NOAA client. I don’t know if that is the only way to get them or not…that function requires an app guid to make the connection/subscription. So, I added a new feature to pull for them by APP GUID during the cache refresh. This feature is switched and can be activated with the cmd line argument --include-container-metrics. We only have one app and it isn’t generating any of these events. I thought it would be something baked into the CF framework as the metrics in question are CPU, MEM, Disk Utilization…but even now that I am attempting to pull them I don’t get any. I will have to investigate back into the core CF code to determine what’s up with that. It may be its WIP (work in progress) or it could be (most likely :)) I am doing something wrong…To support this change I had to refactor the events.go a little to be able to re-utilize the route events func logic. I also renamed the Log func in events to ShipEvent().

I also updated the existing tests to ensure they all still ran after my changes.

Please advise if I can provide any more detail and sorry for the big change set and long story.

Cheers,

Jon
Senior Consultant
ECSTeam Inc.

### CODE REVIEW

File: .project [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (shinji62) [inline .project]: This files in unecessary, you should but into .gitignore

File: firehose/firehose.go [hunk 2]
```diff
@@ -2,25 +2,72 @@ package firehose
...
 
 		for err := range errorChan {
-			fmt.Fprintf(os.Stderr, "%v\n", err.Error())
+			log.LogError("Firehose Error!", err.Error())
 		}
 	}()
 	return msgChan
 }
+
+func GetAppContainerMetrics(DopplerEndpoint string, Token string, app caching.App, skipSSLValidation bool) []*events.Envelope {
```
Comment (shinji62) [inline firehose/firehose.go]: Not sure this one is required... Everything should come from the firehose, let s say we have thousand of apps ...

File: main.go [hunk 3]
```diff
@@ -23,8 +23,9 @@ var (
 	password          = kingpin.Flag("password", "Admin password.").Default("admin").String()
 	skipSSLValidation = kingpin.Flag("skip-ssl-validation", "Please don't").Bool()
 	wantedEvents      = kingpin.Flag("events", fmt.Sprintf("Comma seperated list of events you would like. Valid options are %s", events.GetListAuthorizedEventEvents())).Default("LogMessage").String()
-	boldDatabasePath  = kingpin.Flag("boltdb-path", "Bolt Database path ").Default("my.db").String()
+	boltDatabasePath  = kingpin.Flag("boltdb-path", "Bolt Database path ").Default("my.db").String()
```
Comment (shinji62) [inline main.go]: Thanks for this one :)

**Comment by shinji62** [comment]
Thanks for the PR that's really great, however I have some comment : 

*  firehose-to-syslog would start up and throw an EOF ? 

* As my uaa token would expire it would seem to auto-refresh a few times but eventually I would get a not-authorized error and the executable would abend.
 Token is manage at the go-cfclient using the default oauth2 golang library, so maybe finding the root cause there can be a good idea.

* App Guid that would come through not as a UUID or nil but as a string "<nil>".
I am still wondering how this can be possible, anyway you already put a workaround. 

* As I stated above I was never seeing any ‘Container Metric Events’ come out of the firehose.
My biggest concern come from there, I think we should not lost the logic of only using the firehose endpoint and not created a connexion for everyapp.
As I pointed out https://github.com/cloudfoundry/loggregator/blob/d7b6e85c61da3d7856142dac999c67350f43c5bd/src/integration_tests/doppler/firehose_test.go#L50 firehose should emit container metrics.

Ah you let some unecessary file, like swp, event.test , .project so please remove it :)

And after that please squash your commit :)

**Comment by jonathannewell** [comment]
- EOFs - Please see the below. This is the output & stack trace when I run the original codebase.

macbook-pro:cloudfoundry-community jnewell$ git clone https://github.com/cloudfoundry-community/firehose-to-syslog.git
Cloning into 'firehose-to-syslog'...
remote: Counting objects: 1198, done.
remote: Total 1198 (delta 0), reused 0 (delta 0), pack-reused 1198
Receiving objects: 100% (1198/1198), 955.64 KiB | 1.09 MiB/s, done.
Resolving deltas: 100% (470/470), done.
Checking connectivity... done.
macbook-pro:cloudfoundry-community jnewell$ ls
firehose-to-syslog	go-cfclient
macbook-pro:cloudfoundry-community jnewell$ cd firehose-to-syslog/
macbook-pro:firehose-to-syslog jnewell$ ls
Dockerfile	Godeps		LICENSE		Makefile	README.md	caching		events		firehose	logging		main.go
macbook-pro:firehose-to-syslog jnewell$ go build
macbook-pro:firehose-to-syslog jnewell$ ls -al
total 16872
drwxr-xr-x  15 jnewell  staff      510 Jun  3 08:32 .
drwxr-xr-x   4 jnewell  staff      136 Jun  3 08:32 ..
drwxr-xr-x  13 jnewell  staff      442 Jun  3 08:32 .git
-rw-r--r--   1 jnewell  staff       34 Jun  3 08:32 .gitignore
-rw-r--r--   1 jnewell  staff      128 Jun  3 08:32 Dockerfile
drwxr-xr-x   5 jnewell  staff      170 Jun  3 08:32 Godeps
-rw-r--r--   1 jnewell  staff     1092 Jun  3 08:32 LICENSE
-rw-r--r--   1 jnewell  staff      395 Jun  3 08:32 Makefile
-rw-r--r--   1 jnewell  staff     2993 Jun  3 08:32 README.md
drwxr-xr-x   3 jnewell  staff      102 Jun  3 08:32 caching
drwxr-xr-x   4 jnewell  staff      136 Jun  3 08:32 events
drwxr-xr-x   3 jnewell  staff      102 Jun  3 08:32 firehose
-rwxr-xr-x   1 jnewell  staff  8611028 Jun  3 08:32 firehose-to-syslog
drwxr-xr-x   3 jnewell  staff      102 Jun  3 08:32 logging
-rw-r--r--   1 jnewell  staff     2913 Jun  3 08:32 main.go
macbook-pro:firehose-to-syslog jnewell$ firehose-to-syslog --domain=“#########################” --syslog-server=“#################################:5514" --skip-ssl-validation --user=admin --password=###################### --events=HttpStart,HttpStop,Heartbeat,HttpStartStop,LogMessage,ValueMetric,CounterEvent,ContainerMetric,Error --cc-pull-time=15
2015/06/03 08:34:56 Staring filling app/space/org cache.
2015/06/03 08:34:57 Done filling cache, I will now start processing events!
EOF
EOF
EOF
EOF
EOF
Unauthorized error: You are not authorized. Error: Invalid authorization
Unauthorized error: You are not authorized. Error: Invalid authorization
Unauthorized error: You are not authorized. Error: Invalid authorization
Unauthorized error: You are not authorized. Error: Invalid authorization
panic: close of closed channel

goroutine 15 [running]:
github.com/cloudfoundry-community/firehose-to-syslog/firehose.func·001()
	/Users/jnewell/go/src/github.com/cloudfoundry-community/firehose-to-syslog/firehose/firehose.go:24 +0x314
created by github.com/cloudfoundry-community/firehose-to-syslog/firehose.CreateFirehoseChan
	/Users/jnewell/go/src/github.com/cloudfoundry-community/firehose-to-syslog/firehose/firehose.go:24 +0x31f

goroutine 1 [runnable]:
github.com/cloudfoundry-community/firehose-to-syslog/events.RouteEvents(0xc2080323c0, 0xc2080a7d70)
	/Users/jnewell/go/src/github.com/cloudfoundry-community/firehose-to-syslog/events/events.go:18 +0x6d
main.main()
	/Users/jnewell/go/src/github.com/cloudfoundry-community/firehose-to-syslog/main.go:80 +0x9ea

goroutine 17 [syscall, 10 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 7 [IO wait, 1 minutes]:
net.(*pollDesc).Wait(0xc2080107d0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc2080107d0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc208010770, 0xc2080a0000, 0x1000, 0x1000, 0x0, 0x692d00, 0xc2080b0420)
	/usr/local/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc20803a050, 0xc2080a0000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:121 +0xdc
crypto/tls.(*block).readFromUntil(0xc20803d740, 0x6942f8, 0xc20803a050, 0x5, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:454 +0xe6
crypto/tls.(*Conn).readRecord(0xc20805c000, 0x17, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:539 +0x2da
crypto/tls.(*Conn).Read(0xc20805c000, 0xc20809e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:904 +0x166
net/http.noteEOFReader.Read(0x694700, 0xc20805c000, 0xc2080627e8, 0xc20809e000, 0x1000, 0x1000, 0x32d8c0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1270 +0x6e
net/http.(*noteEOFReader).Read(0xc2080a8400, 0xc20809e000, 0x1000, 0x1000, 0xc208012000, 0x0, 0x0)
	<autogenerated>:125 +0xd4
bufio.(*Reader).fill(0xc208032ae0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Peek(0xc208032ae0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xf0
net/http.(*persistConn).readLoop(0xc208062790)
	/usr/local/go/src/net/http/transport.go:842 +0xa4
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:660 +0xc9f

goroutine 8 [select, 1 minutes]:
net/http.(*persistConn).writeLoop(0xc208062790)
	/usr/local/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:661 +0xcbc

goroutine 10 [chan receive]:
main.func·001()
	/Users/jnewell/go/src/github.com/cloudfoundry-community/firehose-to-syslog/main.go:63 +0x67
created by main.main
	/Users/jnewell/go/src/github.com/cloudfoundry-community/firehose-to-syslog/main.go:68 +0x71e

goroutine 28 [select]:
net/http.(*persistConn).writeLoop(0xc208110000)
	/usr/local/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:661 +0xcbc

goroutine 25 [IO wait]:
net.(*pollDesc).Wait(0xc208065640, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208065640, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc2080655e0, 0xc2080d8000, 0x4000, 0x4000, 0x0, 0x692d00, 0xc2080b0158)
	/usr/local/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc20803a048, 0xc2080d8000, 0x4000, 0x4000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:121 +0xdc
crypto/tls.(*block).readFromUntil(0xc2080a66c0, 0x6942f8, 0xc20803a048, 0x5, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:454 +0xe6
crypto/tls.(*Conn).readRecord(0xc20805cb00, 0x17, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:539 +0x2da
crypto/tls.(*Conn).Read(0xc20805cb00, 0xc2080aa000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:904 +0x166
net/http.noteEOFReader.Read(0x694700, 0xc20805cb00, 0xc2081112e8, 0xc2080aa000, 0x1000, 0x1000, 0x32d8c0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1270 +0x6e
net/http.(*noteEOFReader).Read(0xc2080a9c00, 0xc2080aa000, 0x1000, 0x1000, 0xc208012000, 0x0, 0x0)
	<autogenerated>:125 +0xd4
bufio.(*Reader).fill(0xc208033aa0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Peek(0xc208033aa0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xf0
net/http.(*persistConn).readLoop(0xc208111290)
	/usr/local/go/src/net/http/transport.go:842 +0xa4
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:660 +0xc9f

goroutine 27 [IO wait]:
net.(*pollDesc).Wait(0xc208064220, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208064220, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc2080641c0, 0xc208114000, 0x4000, 0x4000, 0x0, 0x692d00, 0xc208126fa8)
	/usr/local/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc20803a098, 0xc208114000, 0x4000, 0x4000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:121 +0xdc
crypto/tls.(*block).readFromUntil(0xc2080a7080, 0x6942f8, 0xc20803a098, 0x5, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:454 +0xe6
crypto/tls.(*Conn).readRecord(0xc20805cdc0, 0x17, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:539 +0x2da
crypto/tls.(*Conn).Read(0xc20805cdc0, 0xc2080ad000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:904 +0x166
net/http.noteEOFReader.Read(0x694700, 0xc20805cdc0, 0xc208110058, 0xc2080ad000, 0x1000, 0x1000, 0x32d8c0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1270 +0x6e
net/http.(*noteEOFReader).Read(0xc2080a9cc0, 0xc2080ad000, 0x1000, 0x1000, 0xc208012000, 0x0, 0x0)
	<autogenerated>:125 +0xd4
bufio.(*Reader).fill(0xc208033bc0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Peek(0xc208033bc0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xf0
net/http.(*persistConn).readLoop(0xc208110000)
	/usr/local/go/src/net/http/transport.go:842 +0xa4
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:660 +0xc9f

goroutine 26 [select]:
net/http.(*persistConn).writeLoop(0xc208111290)
	/usr/local/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:661 +0xcbc
macbook-pro:firehose-to-syslog jnewell$ 

- files removed and/or added to .gitignore 
- commits squashed.
- removed container metrics code. I am not seeing these metrics come out of the firehose at all. Even running the NOAA firehose sample I don't get any. I am going to have to investigate back into the CF core to see what's up with that.

Please let me know if there is anything else I need to do.

Thanks,

Jon

**Comment by shinji62** [comment]
I think you env is buggy I can get the container metrics 
```json
{"cf_app_id":"c5cb762b-b7bb-44b6-97d1-2b612d4baba9","cf_app_name":"lattice","cf_org_id":"fb5777e6-e234-4832-8844-773114b505b0","cf_org_name":"GWENN","cf_space_id":"3c910823-22e7-41ff-98de-094759594398","cf_space_name":"GWENN-SPACE","cpu_percentage":0.08626019950149455,"disk_bytes":0,"event_type":"ContainerMetric","instance_index":0,"level":"info","memory_bytes":790528,"msg":"","origin":"rep","time":"2015-06-05T22:57:46+09:00"}
{"cf_app_id":"c5cb762b-b7bb-44b6-97d1-2b612d4baba9","cf_app_name":"lattice","cf_org_id":"fb5777e6-e234-4832-8844-773114b505b0","cf_org_name":"GWENN","cf_space_id":"3c910823-22e7-41ff-98de-094759594398","cf_space_name":"GWENN-SPACE","cpu_percentage":0.08679616939340123,"disk_bytes":0,"event_type":"ContainerMetric","instance_index":0,"level":"info","memory_bytes":790528,"msg":"","origin":"rep","time":"2015-06-05T22:58:16+09:00"}
```

For the panic, it will panic if he can connect to the endpoint.

Btw please compile with go 1.4 and using godep

```shell
 godep go build
```

**Comment by jonathannewell** [comment]
ok, its not my environment...it's me. I went through the CF code and no where are any of the components emitting container metrics. I then went back in and looked at the example app in the NOAA client and noticed that it was really two things. 1 to emit metrics to metro and 1 to collect them from the firehose. So, a light finally went on. If I want to see container metrics I have to emit them into the system myself. Then I can collect them. I thought that CF would be emitting container metrics for it's various components, but apparently that is not the plan or has not been implemented yet. Long story short I created a simple go app that uses proc/stats and pulls CPU & MEM metrics and emits them to the local metron (or a configured one) for that vm/instance. Deployed it to the runner vm in my CF install and began getting Container Metrics out of the firehose and into ELK. So, ...just me being dim I guess.  What are next steps to complete the pull request? Cheers, Jon

File: events/events.go [hunk 4]
```diff
@@ -1,65 +1,82 @@
...
+	for _, event := range events {
+		routeEvent(event)
+	}
+}
+
+func routeEvent(msg *events.Envelope) {
+
+	eventType := msg.GetEventType()
+
+	log.LogStd(fmt.Sprintf("Event Type: %s\n", eventType.String()), false)
```
Comment (shinji62) [inline events/events.go]: Really needed ?

File: events/events.go [hunk 5]
```diff
@@ -266,13 +291,23 @@ func ContainerMetric(msg *events.Envelope) Event {
...
-	if cf_app_id != nil && cf_app_id != "" {
-		appInfo := getAppInfo(fmt.Sprintf("%s", cf_app_id))
+	appGuid := ""
+	if cf_app_id != nil {
+		appGuid = fmt.Sprintf("%s", cf_app_id)
+		//fmt.Printf("Annotating for App Id %s\n",appGuid)
+	}
+
+	e.Fields["cf_origin"] = "newell-firehose"
+	e.Fields["cf_event_type"] = e.Type
```
Comment (shinji62) [inline events/events.go]: Why did you change event_type to cf_event_type, that will break dashboard or any logstash filter using event_type.

File: events/events.go [hunk 6]
```diff
@@ -266,13 +291,23 @@ func ContainerMetric(msg *events.Envelope) Event {
...
 	cf_app_id := e.Fields["cf_app_id"]
-	if cf_app_id != nil && cf_app_id != "" {
-		appInfo := getAppInfo(fmt.Sprintf("%s", cf_app_id))
+	appGuid := ""
+	if cf_app_id != nil {
+		appGuid = fmt.Sprintf("%s", cf_app_id)
+		//fmt.Printf("Annotating for App Id %s\n",appGuid)
+	}
+
+	e.Fields["cf_origin"] = "newell-firehose"
```
Comment (shinji62) [inline events/events.go]: Not sure why "newell-firehose", "firehose" should be enough

File: events/events.go [hunk 7]
```diff
@@ -266,13 +291,23 @@ func ContainerMetric(msg *events.Envelope) Event {
...
 
 func (e *Event) AnnotateWithAppData() {
+
 	cf_app_id := e.Fields["cf_app_id"]
-	if cf_app_id != nil && cf_app_id != "" {
-		appInfo := getAppInfo(fmt.Sprintf("%s", cf_app_id))
+	appGuid := ""
+	if cf_app_id != nil {
+		appGuid = fmt.Sprintf("%s", cf_app_id)
+		//fmt.Printf("Annotating for App Id %s\n",appGuid)
```
Comment (shinji62) [inline events/events.go]: can be remove

**Comment by shinji62** [comment]
Last comment added, after that should be good for merging.

**Comment by jonathannewell** [comment]
Ok, I believe I got most of your comments. I am unclear on what you are asking to be removed in the last one. Please advise.

**Comment by shinji62** [comment]
This line can be removed //fmt.Printf("Annotating for App Id %s\n",appGuid)

**Comment by jonathannewell** [comment]
Ok,  I did remove that... :)

**Comment by shinji62** [comment]
Done.
